drivers/rilmodem/gprs-context.c

```diff
@@ -351,8 +394,19 @@ static void ril_deactivate_data_call_cb(struct ril_msg *message,
 				__func__, gcd->apn,
 				ril_error_to_string(message->error));
 
-		if (cb)
+		/*
+		 * It has been detected that some modems fail the deactivation
+		 * temporarily. We do retries to handle that case.
+		 */
+		if (--(gcd->deact_retries) > 0 && gcd->state != STATE_IDLE) {
+			struct cb_data *cbd2;
+
+			cbd2 = cb_data_new(cb, cbd->data, gc);
+			g_timeout_add_seconds(TIME_BETWEEN_DEACT_RETRIES_S,
+							retry_deactivate, cbd2);
+		} else if (cb) {
 			CALLBACK_WITH_FAILURE(cb, cbd->data);
```

Comment:
I think it might be a good idea to add an additional error log message if the retry limit is hit.